app-backend/batch/src/main/scala/com/azavea/rf/batch/ingest/spark/Ingest.scala

```diff
@@ -135,160 +137,90 @@ object Ingest extends SparkJob with LazyLogging with Config {
-        colMax = if (col == chipCols - 1) gb.colMax else col * cw + cw - 1,
-        rowMax = if (row == chipRows - 1) gb.rowMax else row * ch + ch - 1
-      ))
-    }
-  }
-
-  def getSizeFromURI(uri: URI, s3Client: S3Client): Long = {
-    val amazonURI = new AmazonS3URI(uri)
-    val obj = s3Client.getObject(amazonURI.getBucket, amazonURI.getKey)
-    obj.getObjectMetadata.getContentLength
-  }
+    rdd.map { case (_, mbt) => mbt.bands.map(_.histogramDouble(numBuckets)) }
+      .reduce((hs1, hs2) => hs1.zip(hs2).map { case (a, b) => a merge b })
+  // .reduce((hs1, hs2) => (hs1, hs2).parMap2(_ merge _))  // Once Cats 1.0 is released.
 
   /** We need to suppress this warning because there's a perfectly safe `head` call being
     *  made here. The compiler just isn't smart enough to figure that out
     *
     *  @param layer An ingest layer specification
     */
   @SuppressWarnings(Array("TraversableHead"))
-  def ingestLayer(params: CommandLine.Params)(layer: IngestLayer)(implicit sc: SparkContext) = {
+  def ingestLayer(params: CommandLine.Params, layer: IngestLayer)(implicit sc: SparkContext): Unit = {
+
     val resampleMethod = layer.output.resampleMethod
     val tileSize = layer.output.tileSize
     val destCRS = layer.output.crs
     val ndPattern = layer.output.ndPattern
     val bandCount: Int = layer.sources.map(_.bandMaps.map(_.target.index).max).max
     val layoutScheme = ZoomedLayoutScheme(destCRS, tileSize)
-    val s3Client = S3Client.DEFAULT
-    val repartitionSize =
-      layer.sources.map { s =>
-        // Convert partitionsSize from megabytes to bytes
-        math.max(params.partitionsPerFile, getSizeFromURI(s.uri, s3Client) / (params.partitionsSize * 1024 * 1024))
-      }.sum.toInt
-
-    // Read source tiles and reproject them to desired CRS
-    val sourceTiles: RDD[((ProjectedExtent, Int), Tile)] =
-      sc.parallelize(layer.sources, layer.sources.length)
-        .flatMap ({ source =>
-          val geotiff = MultibandGeoTiff(
-            byteReader = uriRangReader(source.uri),
-            decompress = false,
-            streaming = true
-          )
-
-          gridBoundChips(geotiff.tile.gridBounds, params.windowSize, params.windowSize)
-            .map { chipBounds => (source, geotiff.rasterExtent.extentFor(chipBounds)) }
-        })
-        .repartition(repartitionSize)
-        .flatMap { case (source, chipExtent) =>
-          val geotiff = MultibandGeoTiff(
-            byteReader = uriRangReader(source.uri),
-            e = Some(chipExtent)
-          )
-
-          val chip = geotiff.tile
+    val options = S3GeoTiffRDD.Options(maxTileSize = Some(tileSize), partitionBytes = Some(16 * 1024 * 1024))
+
+    val (maxZoom, layerMeta): (Int, TileLayerMetadata[SpatialKey]) =
+      calculateTileLayerMetadata(layer, layoutScheme)
+
+    val rawMultis: Array[RDD[((ProjectedExtent, Int), Tile)]] = layer.sources.map { source =>
+      val uri: AmazonS3URI = new AmazonS3URI(source.uri)
```

Comment:
This line throws an exception when parsing our S3 URIs that I think was handled by logic [here](https://github.com/raster-foundry/raster-foundry/pull/2734/files#diff-b2133f56314df6dce232239b3557e96bL155).

If I change this line to:
`val uri: AmazonS3URI = new AmazonS3URI(source.uri.toString)`

The error goes away and an S3 URI is parsed, but later on I get an error (still trying to figure out if it's related):
```2017-12-22 20:45:47,282 [main] ERROR com.azavea.rf.batch.ingest.spark.Ingest$ -  java.lang.UnsupportedOperationException: empty collection
        at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$apply$36.apply(RDD.scala:1027)
        at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$apply$36.apply(RDD.scala:1027)
        at scala.Option.getOrElse(Option.scala:121)
        at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1027)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
        at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
        at org.apache.spark.rdd.RDD.reduce(RDD.scala:1007)
        at com.azavea.rf.batch.ingest.spark.Ingest$.multibandHistogram(Ingest.scala:141)
        at com.azavea.rf.batch.ingest.spark.Ingest$$anonfun$ingestLayer$1.apply(Ingest.scala:228)
        at com.azavea.rf.batch.ingest.spark.Ingest$$anonfun$ingestLayer$1.apply(Ingest.scala:220)
        at geotrellis.spark.pyramid.Pyramid$.runLevel$1(Pyramid.scala:234)
        at geotrellis.spark.pyramid.Pyramid$.upLevels(Pyramid.scala:242)
        at com.azavea.rf.batch.ingest.spark.Ingest$.ingestLayer(Ingest.scala:220)
        at com.azavea.rf.batch.ingest.spark.Ingest$$anonfun$main$1.apply(Ingest.scala:274)
        at com.azavea.rf.batch.ingest.spark.Ingest$$anonfun$main$1.apply(Ingest.scala:274)
        at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
        at com.azavea.rf.batch.ingest.spark.Ingest$.main(Ingest.scala:274)
        at com.azavea.rf.batch.ingest.spark.Ingest.main(Ingest.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
        at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

```

I'm going to try and move the `tif` somewhere without a broken URL and go with the original parsing and see if that fixes the problem/if it's related.

Reply:
Ok, I verified that removing the pipe `|` from the S3 key allows the current code to run (and the exception above isn't thrown so ingests work). However, I think we need to figure out how to make this backwards compatible as much as that's possible so that the `|` is handled correctly.

Reply:
Alright - think I'm working closer to a resolution here:

`val uri: AmazonS3URI = new AmazonS3URI(java.net.URLDecoder.decode(source.uri.toString, "UTF-8"))` solves most of the problems ... however, in GeoTrellis there is a line in `S3GeoTiffRDD` that does the following:

`uriToKey(new URI(s"s3://${objectRequest.getBucketName}/${objectRequest.getKey}"), k) -> v`

That kind of undermines how `getKey` works in `AmazonS3URI` (since the S3 URI doesn't have issues with `|`) because `getKey` returns the decoded string which might not be a valid URI. I think we will have to add a custom transform to `S3GeoTiffRDD` maybe?